src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -798,45 +813,136 @@ private static void GetForwardedTypes(
+
+                                    foreach (var markerType in groupingType.GetNestedTypes(context))
+                                    {
+                                        builder.Add(new Cci.ExportedType(markerType, groupingIndex, isForwarder: true));
+                                    }
                                 }
+                            }
+                            break;
+
+                        case RetargetingNamedTypeSymbol retargetingType:
+                            {
+                                pushAndProcessNestedTypes(stack, context, index, nested, builder);
+
+                                foreach ((Cci.INestedTypeReference GroupingType, ImmutableArray<Cci.INestedTypeReference> MarkerTypes) item in retargetingType.GetExtensionGroupingAndMarkerTypesForTypeForwarding(context))
+                                {
+                                    int groupingIndex = builder.Count;
+                                    builder.Add(new Cci.ExportedType(item.GroupingType, index, isForwarder: true));
 
-                                stack.Push((nested[i], index));
+                                    foreach (var markerType in item.MarkerTypes)
+                                    {
+                                        builder.Add(new Cci.ExportedType(markerType, groupingIndex, isForwarder: true));
+                                    }
+                                }
                             }
-                        }
+                            break;
+
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(type);
                     }
                 }
+                else
+                {
+                    pushNestedTypes(stack, index, nested);
+                }
+            }
 
-                stack.Free();
+            static void pushNestedTypes(ArrayBuilder<(NamedTypeSymbol type, int parentIndex)> stack, int index, ImmutableArray<NamedTypeSymbol> nested)
+            {
+                // Iterate backwards so they get popped in forward order.
+                for (int i = nested.Length - 1; i >= 0; i--)
+                {
+                    if (nested[i].IsExtension)
+                    {
+                        continue; // Extension blocks are handled separately
+                    }
+
+                    stack.Push((nested[i], index));
+                }
+            }
+
+            static void pushAndProcessNestedTypes(ArrayBuilder<(NamedTypeSymbol type, int parentIndex)> stack, EmitContext context, int index, ImmutableArray<NamedTypeSymbol> nested, ArrayBuilder<Cci.ExportedType> builder)
+            {
+                int currentStackSize = stack.Count;
+
+                pushNestedTypes(stack, index, nested);
+
+                while (stack.Count > currentStackSize)
+                {
+                    processTopItemFromStack(stack, context, builder);
```

Comment:
Why are we processing the newly pushed types here immediately instead of processing them in the top-level loop?